layer.yaml

```diff
@@ -1,31 +1,59 @@
-includes: ['layer:hadoop-client', 'interface:spark', 'interface:benchmark']
+includes:
+- layer:hadoop-client
+- interface:spark
+- interface:benchmark
 options:
-  hadoop-client:
+  apache-spark:
```

Comment:
Thanks for this explanation. I understand where you'd like to go to. I'll
look into the `defines` section. I really like the idea of handing off
common work to lower layers. I wouldn't put the handeling of `dist` in the
hadoop-client layer however. This is functionality that I'd like to use in
other Charms that are not Hadoop clients. It might be better to put this in
a 'dist' layer (or a more descriptive name).

I think the separation between Spark and Livy is very important. First of
all because, i think, it makes code a lot cleaner. Second of all, because
we'll need an approach that can scale. There are a number of similar
projects out there and if the ecosystem starts to grow, so will the number
of additional api's and services. We'll need to be able to handle that
scale, and starting with clear separation between [optional] sub-services
 seems like a step in the right direction.

I'll create a PR to address your concerns. I'd like to keep the separation
of Livy vs Spark vs hadoop-client for now. There isn't much code
duplication anyway because most of the hard work is done by the jujubigdata
module. Although this is something that should be discussed further.



2016-02-29 22:24 GMT+01:00 Cory Johns <notifications@github.com>:

> In layer.yaml
> <https://github.com/juju-solutions/layer-apache-spark/pull/9#discussion_r54480542>
> :
>
> >  options:
> > -  hadoop-client:
> > +  apache-spark:
>
> I didn't see your reply while I was drafting the previous, but I think it
> addresses some of your questions. However, to address explicitly why it
> would belong in hadoop-client, my thought in structuring it that way was
> that it was common functionality that could be provided by the base layer
> (i.e., so that not every client layer had to remember to call
> dist_config.add_users(), etc.), and layer options seemed the most natural
> way to send that data from the charm layer to the base layer.
>
> Layer options was created as a way to provide a well defined, build-time
> validated way to pass data from charm layers to base layers, and to do away
> with the arbitrary yaml files like dist.yaml and apache.yaml.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juju-solutions/layer-apache-spark/pull/9/files#r54480542>
> .
>